extensions/interactions/MathExpressionInput/MathExpressionInput.js

```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+
+            // The focus() call must be in a click event handler and on a text
+            // field to make the mobile keyboard appear.
+            $('#startMathInputButton').click(function() {
+              guppyInstance.activate();
+
+              var fakeInputElement = document.
+                querySelector('#fakeInputForMathExpression');
+              fakeInputElement.focus();
+
+              // Place the cursor at the end of the text input, so that the
+              // user can use backspace to delete.
+              fakeInputElement.setSelectionRange(fakeInputElement.value.length,
+                fakeInputElement.value.length);
+            });
+
+            // Mapping characters back to Mousetrap codes.
+            // Use Mousetrap.trigger(code) to replay them.
+            var K_SYM_REVERSE_MAP = {
+              '^': 'shift+6',
+              '*': 'shift+8',
+              '(': 'shift+9',
+              '<': 'shift+,',
+              '>': 'shift+.',
+              '\\': 'shift+\\',
+              ' ': 'space',
+              ')': 'shift+0'
+              // Missing up/down. user needs to use [space], 'sub' instead.
+              // Mobile devices typically don't have arrow keys anyways.
+            };
+
+            var setGuppyContentFromInput = function() {
+              // Clear the Guppy instance by setting its content to the
+              // output of get_content when empty.
+              guppyInstance.set_content('<m><e></e></m>');
+              guppyInstance.render(true);
+
+              // Get content of the text input field as an array of characters.
+              var textContent = document
+                .querySelector('#fakeInputForMathExpression').value
+                .toLowerCase().split('');
+
+              for (var i = 0; i < textContent.length; i++) {
+                // Replay key combination for each character on the document.
+                if (textContent[i] in K_SYM_REVERSE_MAP) {
+                  Mousetrap.trigger(K_SYM_REVERSE_MAP[textContent[i]]);
+                } else {
+                  Mousetrap.trigger(textContent[i]);
+                }
+              }
+            };
+
+            // Debounce clear/refill cycles to 1 per 100ms.
+            $('#fakeInputForMathExpression').on(
+              'input change compositionupdate keydown',
+              oppiaDebouncer.debounce(function() {
+                setGuppyContentFromInput();
+              }, 100))
+
+            // Set content immediately on blur.
+            .on('blur', function() {
+              guppyInstance.activate();
+              setGuppyContentFromInput();
+            });
+          };
+
           var guppyInstance = new Guppy(guppyDivElt, {
             ready_callback: function() {
               Guppy.get_symbols(
                 GLOBALS.ASSET_DIR_PREFIX +
                 '/assets/overrides/guppy/oppia_symbols.json');
+
+              if (/Mobi/.test(navigator.userAgent)) {
```

Comment:
Break this test out into a separate function -- perhaps put it in deviceInfoService in app.js.